### ISSUE TITLE
Full LavaPlayer format support

### ISSUE THREAD

**Issue by oldmud0** [FlareBot/FlareBot#57]
The implementation would basically be YouTubeExtractor but without the regex URL requirement so that every format supported by LavaPlayer can be supported (WAV, AAC streams, M3U, Ogg streams, SoundCloud, etc). You could probably extract YouTubeExtractor into some sort of superclass, as LavaPlayer takes care of most of the details in handling URLs and there's little need for extra overhead on the FlareBot side of the business. I'd be willing to make a pull request if you're interested.

**Comment by binaryoverload** [comment]
We would be interested in seeing this. If you can make a PR by Friday that is good quality then your implementation might even make it into v3!

**Comment by oldmud0** [comment]
Yay, I'll do it then 😄

**Comment by oldmud0** [comment]
Okay, so this might require some major changes if I really want to have "everything." The reason for this is that there is an unnecessary layer of abstraction between the bot and LavaPlayer, as LavaPlayer is already equipped to parse URLs and such. Using the `DefaultAudioPlayerManager` looks to be significantly simpler than adding an Extractor and fine-grained SourceManagers.

If I kept the current structure, what would end up happening is that everything that isn't YouTube gets funneled into a generic HTTP extractor, and then on top of that, now there is no possibility anymore of selecting all possible SourceManagers, unless I just go with the DefaultPlayerManager which is probably what is desired anyway. However, I cannot use the DefaultPlayerManager because there already exists a pretty sizeable structure of code that overlaps in functionality. On top of that, there is a much simpler way to directly interface with JDA 3 - [see the bottom of the main doc](https://github.com/sedmelluq/lavaplayer#jda-integration).

So, I'll probably need more time to see how I can simplify the music stuff to really unleash the power of LavaPlayer.

**Comment by binaryoverload** [comment]
I wouldn't worry. How if CPU usage likely to be affected by your changes?

**Comment by oldmud0** [comment]
The usage will likely decrease slightly, as LavaPlayer has been specifically optimized for CPU and memory usage. Most of the savings will be from performing the YouTube searches and initiating the media.

**Comment by binaryoverload** [comment]
An initial thing we were concerned about is that, if you add other codec
support, that the CPU usage would go up due to the transcoding

**Comment by oldmud0** [comment]
This might depend chiefly on the bit rate. The best thing to do, I think, would be to prevent loading audio with a bit rate that is too high if we know for sure that the audio is CBR.

**Comment by WalshyDev** [comment]
It doesn't matter about how much bitrate a channel is, we always send the same. We used to support SoundCloud but due to us having to reencode it and this taking up a lot of CPU usage for a feature hardly used it was removed. If it will just be more intensive and not serve any real purpose I don't see the point. I bet if you ask for the stats about what percentage of requests aren't YouTube (or other opus platforms) it will be a lot lower. Is it worth it in your mind to spend more resources providing a feature to a few individuals rather than potentially using those resources for new and cool features every can use?
